### ISSUE TITLE
proposal: Go 2: forbid use unexported types and their aliases as function return types.

### ISSUE THREAD

**Issue by go101** [golang/go#26266]
Go 1 doesn't forbid this, which brings much difficulties for compiler implementations.
It is also a bad unnecessary pattern to do this and it often confuses programmers.

`struct{}` and `interface{}` are two exceptions for this rule.

**Comment by go101** [comment]
sorry, title is mis-described.  Re-edited.

**Comment by mvdan** [comment]
Please explain the rationale in detail. For example, it isn't obvious to me why the "unnamed" part is important, or why struct and interface types are the ones that matter.

For example, what about unnamed struct pointer types? And unnamed struct slices?

**Comment by go101** [comment]
The same. Such types should not be exposed to external packages.

The reason is non-exported fields or methods shouldn't be visible for package users.
Package users can see these fields and methods in package docs, but can't use them, it is confusing.

And using (exported) aliases of non-exported types violates the original intention of the design of non-exported types. (Maybe this should be another proposal)

**Comment by mvdan** [comment]
I agree with the overall sentiment, but hard-coding this in the Go language spec or compiler may not be a good idea. For one, it would likely break lots of existing programs. And this kind of code may still be OK in some edge cases, such as internal packages or generated code.

I would instead propose this as a golint (or even vet) check. That can apply to Go1, and doesn't add to the language spec.
